packages/amplify_core/lib/src/types/models/model_index.dart

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'dart:convert';
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:collection/collection.dart';
+import 'package:flutter/material.dart';
+
+/// Describe an index that is created by `@primaryKey` or `@index` directive,
+/// and is listed in [ModelSchemaDefinition.indexes].
+@immutable
+class ModelIndex {
+  /// Index name that is defined by the name parameter of `@index` directive in
+  /// a model schema. [name] will always be null when the index is representing
+  /// `@primaryKey` directive.
+  final String? name;
+
+  /// An array of field names. The first field is always the field that is
+  /// annotated by` @primaryKey` or `@index` directive, and the remaining
+  /// fields are the fields specified by `sortKeyFields` parameter of
+  /// `@primaryKey` or `@index` directive.
+  final List<String> fields;
+
+  /// Constructor
+  const ModelIndex({
+    required this.fields,
+    this.name,
+  });
+
+  /// Make a copy of an existing [ModelIndex] instance.
+  ModelIndex copyWith({
+    List<String>? fields,
+    String? name,
+  }) =>
+      ModelIndex(fields: fields ?? this.fields, name: name ?? this.name);
+
+  /// Generate a [Map] that represents [ModelIndex].
+  Map<String, dynamic> toMap() {
+    final map = {
+      'name': name,
+      'fields': fields,
+    };
+
+    return map;
+  }
+
+  /// Create an instance of [ModelIndex] from a [Map].
+  factory ModelIndex.fromMap(Map<String, dynamic> map) =>
+      ModelIndex(fields: map['fields'], name: map['name']);
+
+  /// Create an instance of [ModelIndex] from a json
+  factory ModelIndex.fromJson(String source) =>
+      ModelIndex.fromMap(jsonDecode(source));
+
+  /// Generate a json string that represents [ModelIndex]
+  String toJson() => jsonEncode(toMap());
+
+  @override
+  String toString() => 'ModelIndex(name: $name, fields: $fields)';
+
+  @override
+  bool operator ==(Object other) {
```

Comment:
Should we start using the `AWSEquatable` mixin instead of implementing these everywhere? It is currently in amplify_flutter, so you would have to move it if you want to use it. I don't think you have to use it for this PR, but I want to discuss if we want to start using it moving forward.

@dnys1 I think you added AWSEquatable. Any reason we can't move that to amplify_core and start using it throughout all packages?

Reply:
Thanks, wasn't aware of this mixin. Can we also use it with generated models? @dnys1